src/v/security/gssapi_authenticator.cc

```diff
@@ -0,0 +1,514 @@
+        gss::buffer msg;
+        auto maj_stat = gss_display_status(
+          &min_stat, code, type, GSS_C_NO_OID, &msg_ctx, &msg);
+        if (maj_stat != GSS_S_COMPLETE) {
+            vlog(seclog.info, "gss status from {}", m);
+            break;
+        } else {
+            vlog(seclog.info, "GSS_API error {}: {}", m, msg);
+        }
+
+        if (!msg_ctx) {
+            break;
+        }
+    }
+}
+
+std::string display_ctx_flags(OM_uint32 flags) {
+    std::stringstream ss;
+    if (flags & GSS_C_DELEG_FLAG) ss << "context flag: GSS_C_DELEG_FLAG\n";
+    if (flags & GSS_C_MUTUAL_FLAG) ss << "context flag: GSS_C_MUTUAL_FLAG\n";
+    if (flags & GSS_C_REPLAY_FLAG) ss << "context flag: GSS_C_REPLAY_FLAG\n";
+    if (flags & GSS_C_SEQUENCE_FLAG)
+        ss << "context flag: GSS_C_SEQUENCE_FLAG\n";
+    if (flags & GSS_C_CONF_FLAG) ss << "context flag: GSS_C_CONF_FLAG \n";
+    if (flags & GSS_C_INTEG_FLAG) ss << "context flag: GSS_C_INTEG_FLAG \n";
+    return ss.str();
+}
+
+struct hex {
+    explicit hex(bytes const& v)
+      : _v{v} {}
+    explicit hex(bytes_view const& v)
+      : _v{v} {}
+    friend std::ostream& operator<<(std::ostream& os, hex const& h) {
+        for (size_t i = 0; i < h._v.length(); ++i) {
+            if ((i % 16) == 0) {
+                os << '\n';
+            }
+            os << std::setfill('0') << std::setw(2) << std::hex
+               << (int)(h._v[i]) << ' ';
+        }
+        return os;
+    }
+    bytes_view _v;
+};
+
+/*
+ * Function: display_status
+ *
+ * Purpose: displays GSS-API messages
+ *
+ * Arguments:
+ *
+ *      msg             a string to be displayed with the message
+ *      maj_stat        the GSS-API major status code
+ *      min_stat        the GSS-API minor status code
+ *
+ * Effects:
+ *
+ * The GSS-API messages associated with maj_stat and min_stat are
+ * displayed on stderr, each preceeded by "GSS-API error <msg>:
+" and
+ * followed by a newline.
+ */
+void display_status(const char* msg, OM_uint32 maj_stat, OM_uint32 min_stat) {
+    display_status_1(msg, maj_stat, GSS_C_GSS_CODE);
+    display_status_1(msg, min_stat, GSS_C_MECH_CODE);
+}
+
+ss::future<result<bytes>>
+gssapi_authenticator::authenticate(bytes_view auth_bytes) {
+    vlog(seclog.info, "gss {} authenticate", _state);
+    switch (_state) {
+    case state::init: {
+        auto principal = config::shard_local_cfg().sasl_kerberos_principal();
+        auto keytab = config::shard_local_cfg().sasl_kerberos_keytab();
+
+        if (auto res = co_await _worker.submit(
+              [this, principal, keytab]() { return init(principal, keytab); });
+            res.has_error()) {
+            vlog(seclog.info, "gss {} authenticate failed", _state);
+            _state = state::failed;
+            co_return res.assume_error();
+        }
+    }
+        [[fallthrough]];
+    case state::more: {
+        auto res = co_await _worker.submit(
+          [this, auth_bytes]() { return more(auth_bytes); });
```

Comment:
does the worker serialize all operations? i'm thinking about how `more` is a class method and might access shared state?

Reply:
Yeah, they're all backed up on the semaphore.